architecture/spmc.rst

```diff
@@ -0,0 +1,267 @@
+
+Once all SPs are loaded and started we return to the SPMD and the Normal World
+is booted.
+
+
+SP message handling
+-------------------
+
+The SPMC is split into 2 main message handlers:
+
+:``thread_spmc_msg_recv()`` thread_spm.c:	Used to handle message coming
+						from the Normal World.
+:``sp_msg_handler()`` spmc_sp_handler.c:	Used to handle message where
+						the source or the destination
+						is a SP.
+
+When a ``FFA_MSG_SEND_DIRECT_REQ`` message is received by the spmc from the
+Normal World, a new thread is started.
+The FF-A message is passed to the thread and it will call the
+``sp_msg_handler()`` function.
+
+Whenever the SPMC (``sp_msg_handler()``) receives a message not intended for
+one of the SPs, it will exit the thread and return to the Normal World
+passing the FF-A message.
+
+Currently only a ``FFA_MSG_SEND_DIRECT_REQ`` can be passed from the Normal
+World to a SP.
+
+.. uml::
+        :width: 800
+
+        skinparam backgroundcolor transparent
+        participant "None-secure world" as None_secure_world
+
+        box "S-EL3"
+        participant SPMD
+        end box
+
+        box "S-EL1"
+        participant thread_spmc
+        participant spmc_sp_handler
+        end box
+
+
+        box "S-EL0"
+        participant SP
+        end box
+
+        autoactivate on
+        None_secure_world -> SPMD: FFA_MSG_SEND_DIRECT_REQ \n <font color=red>SMC
+        SPMD -> thread_spmc: thread_spmc_msg_recv() \n <font color=red>ERET
+
+        thread_spmc -> spmc_sp_handler : spmc_sp_start_thread()
+        == thread ==
+        spmc_sp_handler -> spmc_sp_handler : sp_msg_handler()
+
+        loop FF-A dst != NSW
+                spmc_sp_handler -> spmc_sp_handler: ffa_handle_sp_direct_req()
+                spmc_sp_handler -> spmc_sp_handler: enter_sp()
+                spmc_sp_handler -> spmc_sp_handler: sp_enter_invoke_cmd()
+                autoactivate off
+                spmc_sp_handler -> SP: __thread_enter_user_mode() \n <font color=red>ERET
+                activate SP
+
+                SP -> spmc_sp_handler : (FF-A message) sp_handle_svc() \n <font color=red>SVC
+                deactivate SP
+                activate spmc_sp_handler
+                spmc_sp_handler -> spmc_sp_handler: Store SP context
+                spmc_sp_handler -> spmc_sp_handler: return to sp_enter_invoke_cmd()
+                deactivate spmc_sp_handler
+                spmc_sp_handler -> spmc_sp_handler: Retrieve FF-A message from SP context
+                return
+                return
+                return
+        end
+                return
+        == End of thread ==
+                return
+        return \n <font color=red>SMC
+        return FFA_MSG_SEND_DIRECT_RESP \n <font color=red>ERET
+
+
+Every message received by the SPMC from the Normal World is handled in the
+``thread_spmc_msg_recv()`` function.
+
+When entering a SPs we need to be running on a thread. This is needed to be
```

Comment:
s/SPs/SP/
s/on a thread/in an OP-TEE thread/

Reply:
Done